### DISCUSSION TITLE
[SP-API][Finances API Testing] 403 Unauthorized errors when trying to test Finances API v0 via Postman

### DISCUSSION THREAD

**Discussion by carlshi12r44** [amzn/selling-partner-api-models#3086] [General]
Hello team,
  we are trying to test out SP-API with `listFinancialEvents` operation (on this [Finances API references](https://developer-docs.amazon.com/sp-api/docs/finances-api-reference)) in postman, but encountered `403 - Unauthorized - Access to requested resource is denied` error when we tried to call the endpoints in Postman.

### What we had done/How to reproduce?
Mainly followed the [IAM configure doc](https://developer-docs.amazon.com/sp-api/docs/creating-and-configuring-iam-policies-and-entities), [Register application doc](https://developer-docs.amazon.com/sp-api/docs/registering-your-application), and [Making request to SP-API doc](https://developer-docs.amazon.com/sp-api/docs/using-postman-for-selling-partner-api-models)
1. We setup the AWS IAM user, IAM policy for execute-api , STS policy, and IAM role with one of our AWS account (different than the seller account we had)
2. We registered our application and the status is in `Draft`, and we chose `
Finance and Accounting` as our only role, and we verified that it could permit operation `listFinancialEvents` 
3. We obtained the `refresh_token` by choosing `authorize` in the `Action` drop down in the Developer Central page. 
4. We also obtained the LWA credentials from application page in develop central page.
5. We made an access token (started with `Atza|`) request with LWA credentials and the refresh tokens in postman, and we successfully obtained the access_token.
6. We also made a STS request with our IAM role ARN and successfully obtained temporary credentials for `access key` and `secret key`, and `session token` from the STS request.
7. We put the `access key` and `secret key`, and `session token` from step 6 to the authorization tab in the listFinancialEvents request in postman, and we also copied and pasted the access token from step 5 to `x-amz-access-token` in the header tab in postman
### what was the issue?
we were getting `403 Unauthorized - Access to requested resource is denied` error when we sent the request
```
{
    "errors": [
        {
            "message": "Access to requested resource is denied.",
            "code": "Unauthorized",
            "details": ""
        }
    ]
}
```

Our hypothesis is that we didn't go to through the [Authorizing Selling Partner API applications](https://developer-docs.amazon.com/sp-api/docs/authorizing-selling-partner-api-applications) due to lack of context here: we thought getting a refresh token would be sufficient to generate temporary credentials to sign when we made requests to the endpoints, so we left OAuth authorization URI empty in the app settings. Should we add the OAuth URI as instructed in the guide? If not, what could be other causes?

We raised a support case with caseID = `11191766901`

Thank you

**Comment by ritikacd** [comment]
Hi carlshi12r44,

If you have verified that you have all of the correct roles and permissions, check that your IAM ARN is added to the application. The IAM ARN you add should be the same IAM ARN to which the IAM policy was attached during registration process.

a) If IAM ARN is associated with an IAM role, make sure that the IAM policy is attached to the role. If this is case, only then you request for the temporary credentials from AWS STS and pass the credentials in SP-API endpoints.
b) If IAM ARN is associated with an IAM user, make sure that the IAM policy is attached to the IAM user and not to the IAM role. If this is the case, then skip getting temporary credentials from AWS STS and directly use AWS access key ID and AWS secret access key obtained during IAM ARN configuration for the IAM user. 

We have a lovely blog on troubleshooting SP-API errors you may want to explore: https://developer-docs.amazon.com/sp-api-blog/docs/troubleshooting-selling-partner-api-authorization-errors

Thanks,
Ritika

**Reply by carlshi12r44** [reply]
Hi Ritika,
  thanks for the information and support provided, so you mean it would not work if I have attached IAM policy to both IAM role & IAM user? Only attach the policy to either one would be a solution? I have tried with a different seller account and authorized it with new IAM user but still no luck.

**Reply by ritikacd** [reply]
Hi carlshi12r44,
It depends on the IAM ARN provided while registering your application on the Developer Central page, can you check if you have IAM user or role ARN specified "IAM ARN" field? 

Thanks,
Ritika

**Reply by carlshi12r44** [reply]
HI Ritika,
  for my current application, I specified IAM user in the "IAM ARN" field, I am also confirmed that IAM user has IAM execute-api inline policy attached, and its ARN on the app matches with the one listed in the IAM user in my AWS console.

**Reply by carlshi12r44** [reply]
Another question I am wanna to make sure is the the developer type. We just want to go through self-authorization route instead of getting authorized by a seller. Should we change our developer type to `private seller`? Thank you

**Reply by ritikacd** [reply]
No need to change the developer type, you can self authorize a public application as well. 

In the listFinancialEvents postman request, under Authorization tab replace AccessKey and SecretKey obtained from AWS(it's the same Access key pair used to generate temporary credentials using AWS STS.)

If your application is associated with IAM user ARN you will need to use Access keys obtained in [Step 2 Create an IAM user](https://developer-docs.amazon.com/sp-api/docs/creating-and-configuring-iam-policies-and-entities#step-2-create-an-iam-user)

Postman request Authorization tab sample(Session Token stays empty):
![image](https://user-images.githubusercontent.com/108386887/201018448-aa6d4b0b-dec1-4cdb-a4b2-96be40a06a21.png)

Hope this helps!

**Reply by carlshi12r44** [reply]
Hi I followed the step but still no luck.

**Reply by ritikacd** [reply]
Hi carlshi, please share screenshot of the exact steps followed

**Reply by carlshi12r44** [reply]
yes here's are the screenshots, I used another app with IAM role
STS request param tab
![Screen Shot 2022-11-10 at 12 06 06 PM](https://user-images.githubusercontent.com/20281318/201196777-cf7a0967-b633-4429-a180-3c8b20d70018.png)

STS request Auth tab
![Screen Shot 2022-11-10 at 12 09 02 PM](https://user-images.githubusercontent.com/20281318/201196847-a90dc1ce-7481-4a97-9a4c-a08a9852f59d.png)

Access token Auth tab
![Screen Shot 2022-11-10 at 12 04 37 PM](https://user-images.githubusercontent.com/20281318/201196166-90c237bf-84b6-40be-81c0-4fa56873ffa9.png)

Access token Body tab
![Screen Shot 2022-11-10 at 12 12 59 PM](https://user-images.githubusercontent.com/20281318/201196728-f17abb32-69f2-4693-99dd-ebd527c30c9e.png)

ListFinancialEvents Request Auth tab
![Screen Shot 2022-11-10 at 12 09 55 PM](https://user-images.githubusercontent.com/20281318/201196921-c0cea2ff-90e9-4d5c-b610-63d08c5a52a0.png)


ListFinancialEvents Request Headers tab
![Screen Shot 2022-11-10 at 12 10 18 PM](https://user-images.githubusercontent.com/20281318/201196967-e08f4846-e236-419d-941c-208d0e194c15.png)

**Reply by ritikacd** [reply]
The above steps look right if your application is associated with the IAM role ARN. We are unable to view the endpoint used, can you check the seller account that you are making request to is in the same region as the endpoint you are using in the request? I assume it would be North America region as AWS region used in the requests is "us-east-1". 

If you still see issues, please reach out to Developer [Support](https://developer.amazonservices.com/support) so they can investigate the issue further. Support has access to request IDs to debug the issue. 

Thanks!

**Reply by carlshi12r44** [reply]
Hi ritikacd,
  thanks, the error persists but some one from the the support team is currently checking it. Thank you so much for your help:)

Best,
Carl

**Reply by sungolivia** [reply]
keep us updated @carlshi12r44  this issue comes up a lot amongst our devs out there!

**Reply by carlshi12r44** [reply]
Hi there, SP API support team helped me with this issue by raising a support case in the Seller Central.

**Reply by sungolivia** [reply]
was there anything specific issue that you can share the resolution of? if it's specific to your account of course no need to share!

**Comment by tenbasetom** [comment]
I just submitted a new ticket that could be related. Also getting 403 :( -- https://github.com/amzn/selling-partner-api-docs/issues/3048

We're seeing this a lot with the new API.
